## Numpy assignment

## 1 Preliminaries

In [1]:
# The usual preamble
import numpy as np

## Numpy

In [2]:
import numpy as np
x = np.array([2,3,1,0])
y =np.arange(10)
z = 2 + x

In [3]:
x, y, z

(array([2, 3, 1, 0]),
 array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
 array([4, 5, 3, 2]))

Write expressions in the next cell to retrieve 1 from `x`, 3 from `y`, and 3 from `z`.

In [4]:
print(x[2])
print(y[3])
print(z[2])

1
3
3


In the next cell, write an expression that generates a 5 by 5 array filled with zeros, and another  that generates a 1 by 1 array with a 1 in it.

In [5]:
five_by_five = np.zeros((5,5))
one_by_one = np.ones((1,1))

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  

In the next cell, write an expression that uses an assignment to a splice to make all the negative values in `a` be 0.  Attention: This can be done more easily in numpy than it can in normal Python.  See if you can do it the easy way.

In [8]:
a = np.arange(10,-1,-1) - 3
print(a)
a[-3:] = 0
print(a)

[ 7  6  5  4  3  2  1  0 -1 -2 -3]
[7 6 5 4 3 2 1 0 0 0 0]


In the next cell write an expression that produces an array containing result of multiplying 7 by each of the first 12 integers (1 - 12). There's a hard way to do this and an easy way.  The easy way uses elementwise operations.

In [9]:
7 * np.arange(1,13)

array([ 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84])

## 2D arrays

In [10]:
import numpy as np
room_matrix = \
np.array(
[[6,  3, 4, 1],
[5,  2, 3, 2],
[8,  3, 6, 2],
[5,  1, 3, 1],
[10, 4, 7, 2]])

cost_vector = np.array([40, 175, 90, 450])

In the next cell write an express that retrieves the last column of `room_matrix`.

In [11]:
room_matrix[:,3]

array([1, 2, 2, 1, 2])

In the next cell, write an expression that retrieves the third row of `room_matrix`.

In [12]:
room_matrix[2,:]

array([8, 3, 6, 2])

In the next cell write an expression that retrieves the first two rows of the second column of `room_matrix`

In [13]:
room_matrix[0:2,1]

array([3, 2])

In the next cell write an expression that retrieves the following submatrix from `room_matrix`:

```
array([[2,3],
       [3,6]])
```

In [14]:
room_matrix[1:3,1:3]

array([[2, 3],
       [3, 6]])

In the next cell, write an expression that returns a Boolean array that identifies the members of `room_matrix` which are greater than or equal to 7.

In [15]:
room_matrix >= 7

array([[False, False, False, False],
       [False, False, False, False],
       [ True, False, False, False],
       [False, False, False, False],
       [ True, False,  True, False]])

## Iris data exercise

In the next cell, use a Boolean mask to return all the rows of the iris data set whose column 3 value is greater than 1.8 (I mean the column whose **index** is 3, the last column).
The cell is not empty because you have been provided with the code which loads the data.
So you should keep that code and **add to it**.

In [16]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']
features[features[:,3]>1.8]

array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.7, 3.3, 5.7, 2.1],
       [6.4, 2.8, 5.6, 2.1],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [7.7, 3. , 6.1, 2.3],
       [6.3, 3.4, 5.6, 2.4],
       [6.9, 3.1, 5.4, 2.1],
       [6.7, 3.1, 5.6, 2.4],
       [6.9, 3.1, 5.1, 2.3],
       [5.8, 2.7, 5.1, 1.9],
       [6.8, 3.2, 5.9, 2.3],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 3. , 5.2, 2.3],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3]])

In the next cell, use a Boolean mask to return all the members of class 2 in the iris data set.

In [ ]:
features[target == 2,:]

Extra credit:  In the next cell write an expression -- or expressions -- which returns
all the members of class 2 in the data whose column 3 value is greater than 1.8.

Hint: Set a variable to store the result of one of your answers to the last two questions.
That result is an array, and you can do more indexing on it.

In [6]:
cls2 = features[target == 2,:]
cls2[cls2[:,3]>1.8]

array([[ 6.3,  3.3,  6. ,  2.5],
       [ 5.8,  2.7,  5.1,  1.9],
       [ 7.1,  3. ,  5.9,  2.1],
       [ 6.5,  3. ,  5.8,  2.2],
       [ 7.6,  3. ,  6.6,  2.1],
       [ 7.2,  3.6,  6.1,  2.5],
       [ 6.5,  3.2,  5.1,  2. ],
       [ 6.4,  2.7,  5.3,  1.9],
       [ 6.8,  3. ,  5.5,  2.1],
       [ 5.7,  2.5,  5. ,  2. ],
       [ 5.8,  2.8,  5.1,  2.4],
       [ 6.4,  3.2,  5.3,  2.3],
       [ 7.7,  3.8,  6.7,  2.2],
       [ 7.7,  2.6,  6.9,  2.3],
       [ 6.9,  3.2,  5.7,  2.3],
       [ 5.6,  2.8,  4.9,  2. ],
       [ 7.7,  2.8,  6.7,  2. ],
       [ 6.7,  3.3,  5.7,  2.1],
       [ 6.4,  2.8,  5.6,  2.1],
       [ 7.4,  2.8,  6.1,  1.9],
       [ 7.9,  3.8,  6.4,  2. ],
       [ 6.4,  2.8,  5.6,  2.2],
       [ 7.7,  3. ,  6.1,  2.3],
       [ 6.3,  3.4,  5.6,  2.4],
       [ 6.9,  3.1,  5.4,  2.1],
       [ 6.7,  3.1,  5.6,  2.4],
       [ 6.9,  3.1,  5.1,  2.3],
       [ 5.8,  2.7,  5.1,  1.9],
       [ 6.8,  3.2,  5.9,  2.3],
       [ 6.7,  3.3,  5.7,  2.5],
       [ 6

Note it doesn't work to do this in the other order:

In [7]:
large_col_three = features[features[:,3]>1.8]
large_col_three[target==2,:]

/Users/gawron/Library/Enthought/Canopy_64bit/User/lib/python2.7/site-packages/ipykernel/__main__.py:2: VisibleDeprecationWarning: boolean index did not match indexed array along dimension 0; dimension is 34 but corresponding boolean dimension is 150
  from ipykernel import kernelapp as app


IndexError: index 100 is out of bounds for axis 0 with size 34

This is because the Boolean array used for indexing must be exactly the same length as the array indexed:

```
indexed[indexing == 'something']
```
`indexing == 'something'` needs to return an array of Booleans of exactly the same length
as `indexed`.  For that to happen, `indexing` has to be the same length as `indexed`.
This is analogous to the constraint on elementwise arithmetic between
arrays. If you add or multiply two arrays `X` and `Y`, they need to be the same length,
too.  The array `target` is the same length as `features`, so

```
features[target==2,:]
```
works. You get back the rows of `features` that correspond to a `True` in the Boolean vector
`target == 2`.  But when you do

```
large_col_three = features[features[:,3]>1.8]
```

you get an array of reduced size, which is no longer the same length as `target`.
Hence

```
large_col_three[target==2,:]
```

raises an error.

## Boston data exercise

In [17]:
from sklearn import datasets
boston = datasets.load_boston()
X,Y = boston.data,boston.target
print(X.shape, Y.shape)

(506, 13) (506,)


In [4]:
X[:,3]

array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  0.,  1.,  1.,
        0.,  0.,  0.,  0.,  1.,  0.,  1.,  1.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0

The cell above loads the `sklearn` Boston housing data, which contains information about housing prices
for 501 houses sold in Boston, and the values of various attributes which have been found to be useful
in predicting home prices:

```
Index
0         - CRIM     per capita crime rate by town
1         - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
2         - INDUS    proportion of non-retail business acres per town
3         - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
4         - NOX      nitric oxides concentration (parts per 10 million)
5         - RM       average number of rooms per dwelling
6         - AGE      proportion of owner-occupied units built prior to 1940
7         - DIS      weighted distances to five Boston employment centres
8         - RAD      index of accessibility to radial highways
9         - TAX      full-value property-tax rate per $10,000
10        - PTRATIO  pupil-teacher ratio by town
11        - B        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
12        - LSTAT    % lower status of the population
```

The array `X`, defined in the cell above, is a 501x13 array containing the values of all
these variables for 501 Boston houses.  

The information stored in Y, the value to be predicted is

```
       - MEDV     Median value of owner-occupied homes in $1000's
 
```

1.  Construct an expression or expressions that creates a new table that omits all houses whose tracts
    bound the Charles River.
2.  Construct an expression or expressions that creates a new table containing only the CRIM column and
    the ZN column.  This is  our new X.  Also craete a new Y containing the pupil-teach ratio.  
    This is to help someone research how to predict poor education conditions.  The shape of the new X should
    (506,2) and the shape of the new Y should be (506,).  Show that you checked this in your answer.

Solution to 1.

In [9]:
(X[:,3]==0).shape

(506,)

In [7]:
NewX = X[X[:,3]==0,:]

In [8]:
print(X.shape)
print(NewX.shape)

(506, 13)
(471, 13)


Solution to 2

In [14]:
NewX = X[:,0:2]
NewY = X[:,10]
NewX.shape, NewY.shape

((506, 2), (506,))

## Using numpy arrays for frequency data

In this section we build a 2D array representing a textual data set. There are tools that will do what we doing here much faster than the code you write, but the idea is for you to gain a better understanding of what the actual computational representation of the data is by building it yourself.

Steps

1.  We loop through a set of documents getting the vocab counts for each.  Our final vocabulary is the union of the vocabularies of the documents, i.e., every word we've seen in the data set, even if it occurred in only one document.
2. Let D be the number of documents.  Let V be the vocab size.  We build a DxV matrix M representing the frequency counts of each word in each document.  `M[i,j]` is the count of the `j`-th word in the `i`-th document.

The matrix M can be passed directly to a machine learning algorithm as its training matrix.


In [2]:
d = 'northanger_abbey'
fh = open(f'austen/{d}.txt','r')

In [7]:
text = fh.readlines()

In [8]:
len(text)

8249

In [10]:
fh.readlines()

[]

In [1]:
import numpy as np
import numpy.linalg as LA
from collections import Counter

docset = ['pride_and_prejudice','northanger_abbey',
          'emma', 'mansfield_park',
          'sense_and_sensibility', 'persuasion']
vocab = set()
doc_counts = []
for d in docset:
    with open(f'austen/{d}.txt','r') as fh:
        ctr = Counter(fh.read().lower().split())
        vocab.update(list(ctr.keys()))
        doc_counts.append(ctr)

vocab = sorted(list(vocab))
(D,V) = (len(doc_counts), len(vocab))

M = np.zeros((D,V))
for i in range(D):
    doc_ctr_i = doc_counts[i]
    for j in range(V):
        M[i,j] = doc_ctr_i[vocab[j]]

# The words counts for each document.
# A vector of length 6. Applying norm
# function to rows, M[0,:], M[1,:], M[2,:], etc., 
# i.e., slices along dimension 1
doc_sizes = LA.norm(M,ord=1,axis=1)

#def divide (x,y):
#    return x/y

# Divide every word count in a doc by the doc size of the doc
# Divide each column vector [a vector of length 6] elementwise
# by doc_sizes [another vector of length 6].
# We are applying the function to columns, 
# M[:,0],M[:,1,],M[:,2], etc., i.e., slices along dimension, 0.
M_norm = np.apply_along_axis(np.divide,0,M,doc_sizes)

Understanding the `apply_along_axis` function.  The basic idea is to apply any function of a 1D array, to either rows or columns of a matrix:

In [23]:
def my_average(a):
    """Find size of average element of a 1-D array"""
    return sum(a)/float(len(a))

# a 3x3 array.
#X = np.array([[1,2,3,4,5], [4,5,6,7,8], [7,8,9,10,11]])
X = np.array([[1,2,3,4,5], [6,7,8,9,10], [11,12,13,14,15]])
X

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15]])

In [46]:
print('Axis 0, apply my_average to 5 cols of X',np.apply_along_axis(my_average, 0, X))
print('Axis 1, apply my_average to to 3 rows of X',np.apply_along_axis(my_average, 1, X))

Axis 0, apply my_average to 5 cols of X [  6.   7.   8.   9.  10.]
Axis 1, apply my_average to to 3 rows of X [  3.   8.  13.]


In computing `M_norm` above, we gave `apply_along_axis` one more argument `doc_sizes` which is interpreted as an additional argument of `np.divide`, so this means we divide every column elementwise by the document sizes vector.

Note that `M` gives the counts for a word in a document, and `M_norm` the proportion of the document the word occupies.

In [25]:
print(M.shape, M_norm.shape)
print(docset.index('northanger_abbey'), docset[1])
print(docset.index('pride_and_prejudice'),docset[0])
print(vocab.index('the'),vocab[34915])

print(M[1,34915], M_norm[1,34915])
print(M[0,34915], M_norm[0,34915])

(6, 39548) (6, 39548)
1 northanger_abbey
0 pride_and_prejudice
34915 the
3321.0 0.0414327419717
4479.0 0.0359504928243


Read the code above carefully, make sure you understand it. Answer the following questions about `M` and `M_norm`.

1)  What is the size of Jane Austen's vocabulary?  How many words long is the Jane Austen canon (the sum of the lengths [in words] of the 6 novels)?

In [48]:
print('Vocab size', len(vocab))
print('Size of canon', saum(doc_sizes))

Vocab size 39548
Size of canon 735640.0


In [2]:
doc_sizes

array([ 124588.,   80154.,  160449.,  162553.,  121590.,   86306.])

No, she has not passed the million word milestone.

2)  How many times does "time" occur in *Pride and Prejudice*?  Show the computations you use to compute this. 

In [47]:
time_index = vocab.index('time')
pride_and_prejudice_index = docset.index('pride_and_prejudice')
print(time_index)
print(pride_and_prejudice_index) 
print('Count of "time" in "Pride and Prejudice":',M[pride_and_prejudice_index,time_index])

35427
0
Count of "time" in "Pride and Prejudice": 136.0


3)  How many times does "year" occur in *Sense and Sensibility*?  Show your work.

In [49]:
year_index = vocab.index('year')
sense_and_sensibility_index = docset.index('sense_and_sensibility')
print(year_index)
print(sense_and_sensibility_index)
print('Count of "year" in "Sense and Sensibility":',M[sense_and_sensibility_index,year_index])

39291
4
Count of "year" in "Sense and Sensibility": 14.0


4)  Note that 'It' is not present in the vocabulary.  Can we conclude that Jane Austen never starts a sentence with 'It'?  Why or why not?

The entire string of each novel is lower cased before being passed to the counter for that novel:

```
ctr = Counter(fh.read().lower().split())
```

So we only know that the word `it` occurred, not whether it was lower or upper case.

5)  Compute an array which gives the counts for the word "gentleman" in all 6 Jane Austen novels.

In [50]:
gentleman_index = vocab.index('gentleman')
M[:,gentleman_index]

array([ 18.,  20.,  14.,  10.,  18.,  12.])

We see that one novel has 20 occurrences of `gentleman` in the count vector above.  That is the novel with index 1. To see which one that is, we do:

In [33]:
docset[1]

'northanger_abbey'

In two lines of code this is:

In [51]:
gentleman_counts = list(M[:,gentleman_index])
docset[gentleman_counts.index(max(gentleman_counts))]

'northanger_abbey'

6)  Which novel has the largest number of tokens of "truth"?  Show your computations.

In [52]:
truth_index = vocab.index('truth')
truth_counts = list(M[:,truth_index])
docset[truth_counts.index(max(truth_counts))]

'emma'

7)  Which novel has the largest proportion of occurrences of "lady" and which the largest proportion of occurrences of "gentleman"?  Which word shows up more often in Austen's writings?

In [44]:
lady_index = vocab.index('lady')
lady_props = list(M_norm[:,lady_index])
print('Greatest proportion of tokens of "lady":', docset[lady_props.index(max(lady_props))])

Greatest proportion of tokens of "lady": persuasion


In [45]:
gentleman_index = vocab.index('gentleman')
gentleman_counts = list(M_norm[:,gentleman_index])
print('Greatest proportion of tokens of "gentleman":', docset[gentleman_counts.index(max(gentleman_counts))])

Greatest proportion of tokens of "gentleman": northanger_abbey


In [56]:
lady_index = vocab.index('lady')
gentleman_index = vocab.index('gentleman')
total_lady_counts = sum(M[:,lady_index])
total_gentleman_counts = sum(M[:,gentleman_index])
print('Total counts  "lady"', total_lady_counts)
print('Total counts  "gentleman"', total_gentleman_counts)

Total counts  "lady" 716.0
Total counts  "gentleman" 92.0


Note: It's not even close.

In [57]:
print(M[:,lady_index])
print(M[:,gentleman_index])

[ 166.   27.   32.  169.  122.  200.]
[ 18.  20.  14.  10.  18.  12.]


*Northanger Abbey* and *Emma* are the outliers for *lady*.

Temporal order:

1. Sense and Sensibility
2. Pride and Prejudice
3. Northanger Abbey
4. Mansfield Park
5. Emma
6. Persuasion
